### PR TITLE
Static event handlers inference

### PR DESCRIPTION
**Pull request by itaipelles**

### CODE REVIEW

**Comment by ohana54** [comment]
I didn't have time to fully review, but 2 things:
1. What happens when an element is deleted but the event handler remains. 
2. The whole code is very hard to read and understand, I think we need better abstractions on top of tern. 
What do you think?

**Comment by itaipelles** [comment]
1. the events data we get from santa-editor is represented as a list of components, each with a .events property. so we can not programmatically receive events with no matching component. Event handlers that are abscent from the symbols will simply exit the function here https://github.com/wix/wix-code-ide/blob/72ebf473e5557ecd075803861bb5eeb6da83497a/src/wix-code/tern/plugins/wixStaticEventHandlers.js#L92 and won't get any type inference.

2. I did a pretty big refactor and i think it's much more readable now. Please give it another try and let's discuss after that? I think wrapping tern with an abstraction layer that makes sense is a very good idea, but i'm not sure we're ready for it yet.

**Comment by itaipelles** [comment]
@ob1 
I implemented the changes we discussed and did some refactoring.
Also, with Sam pushing the documentation of the new $w parameter to all events, the previous solution did not exactly work.. So now i'm putting the types data of the parameters on the scope object, rather than on the function type object (why is there no single source of truth..? :( )
I also added tests for $w in static event handlers.
Can you have a look?

File: src/wix-code/tern/plugins/wixStaticEventHandlers.js [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+    function getEventHandlerType(symbols, callbackSymbol) {
+        var compSdkType = symbols.idTable[callbackSymbol.componentName].type;
+        var compTernType = ternUtils.getTernTypeBySdkType(infer, wixCodeTern, compSdkType);
+        var eventPropName = componentEventTypeToPropName[compSdkType.shortName][callbackSymbol.eventType];
+        return ternUtils.safeGet(compTernType, ['proto', 'props',eventPropName, 'types', '0', 'args', '0']);
+    }
+
+    function putArgumentTypesOnMatchingPropertiesOfScope(actualHandlerFunction, ast, eventHandlerType) {
+        var actualPropNames = actualHandlerFunction.types[0].argNames;
```
Comment (ohana54) [inline src/wix-code/tern/plugins/wixStaticEventHandlers.js]: `functionArgumentNamesInCode`

File: src/wix-code/tern/plugins/wixStaticEventHandlers.js [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        var functionScope = infer.scopeAt(ast, actualHandlerFunction.originNode.end);
```
Comment (ohana54) [inline src/wix-code/tern/plugins/wixStaticEventHandlers.js]: `functionBodyScope`?

File: src/wix-code/tern/plugins/wixStaticEventHandlers.js [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        Object.keys(functionScope.props).forEach(function(functionScopePropName) {
```
Comment (ohana54) [inline src/wix-code/tern/plugins/wixStaticEventHandlers.js]: refactor to run on the actual prop names

File: src/wix-code/tern/plugins/wixStaticEventHandlers.js [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+            'wix-code/utils/selectorParser',
+            'wix-code/tern/ternUtils',
+            'wix-code/tern/definitions/wixCodeTern',
+            'wix-code/tern/plugins/symbolService'], mod);
+    mod(infer, tern, cst, selectorParser, ternUtils, wixCodeTern, symbolService);
+})(function(infer, tern, cst, selectorParser, ternUtils, wixCodeTern, symbolService) {
+
+    var componentEventTypeToPropName = null;
+
+    function postLoadDef() {
```
Comment (ohana54) [inline src/wix-code/tern/plugins/wixStaticEventHandlers.js]: Remove and do it when the module loads

File: src/wix-code/tern/plugins/wixStaticEventHandlers.js [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        componentEventTypeToPropName = {};
+        var $wNameSpace = wixCodeTern['!define'].$w;
+        var componentTypes = Object.keys($wNameSpace);
+        componentTypes.forEach(function(componentType) {
+            var prototype = $wNameSpace[componentType].prototype;
+            if (!prototype) {
```
Comment (ohana54) [inline src/wix-code/tern/plugins/wixStaticEventHandlers.js]: why? write a comment with an example

File: test/browser/autocomplete/event-callbacks.spec.js [hunk 6]
```diff
@@ -18,29 +18,85 @@ describe('event callbacks', () => {
         await this.browser.get(`${this.baseUrl}/test/playground/index.html`);
     });
 
-    it('should display proposals for MouseEvent', async function() {
-        const editorLoaderDriver = editorLoaderDriverCreator(this.browser);
+    describe('dynamic', function() {
```
Comment (ohana54) [inline test/browser/autocomplete/event-callbacks.spec.js]: handle all possible cases

**Pull request merged**